common/beerocks/bcl/include/bcl/network/sockets_impl.h

```diff
@@ -0,0 +1,348 @@
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    socklen_t length() const override { return m_length; }
+
+private:
+    sockaddr_un m_address = {};
+    socklen_t m_length    = sizeof(m_address);
+};
+
+class InternetAddress : public Socket::Address {
+public:
+    explicit InternetAddress(uint16_t port, uint32_t address = INADDR_ANY)
+    {
+        m_address.sin_family      = AF_INET;
+        m_address.sin_addr.s_addr = address;
+        m_address.sin_port        = htons(port);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    socklen_t length() const override { return m_length; }
+
+private:
+    sockaddr_in m_address = {};
+    socklen_t m_length    = sizeof(m_address);
+};
+
+class LinkLevelAddress : public Socket::Address {
+public:
+    LinkLevelAddress() {}
+
+    LinkLevelAddress(uint32_t iface_index, const sMacAddr &mac)
+    {
+        m_address.sll_family  = AF_PACKET;
+        m_address.sll_ifindex = iface_index;
+        m_address.sll_halen   = sizeof(sMacAddr);
+        std::copy_n(mac.oct, sizeof(sMacAddr), m_address.sll_addr);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    socklen_t length() const override { return m_length; }
+
+private:
+    sockaddr_ll m_address = {};
+    socklen_t m_length    = sizeof(m_address);
+};
+
+class NetlinkAddress : public Socket::Address {
+public:
+    explicit NetlinkAddress(uint32_t groups = 0)
+    {
+        m_address.nl_family = AF_NETLINK;
+        m_address.nl_groups = groups;
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    socklen_t length() const override { return m_length; }
+
+private:
+    sockaddr_nl m_address = {};
+    socklen_t m_length    = sizeof(m_address);
+};
+
+/**
+ * This class is a wrapper for the socket file descriptor obtained with the accept() system call.
+ */
+class ConnectedSocket : public SocketAbstractImpl {
+public:
+    explicit ConnectedSocket(int fd) : SocketAbstractImpl(fd) {}
+};
+
+class RawSocket : public SocketAbstractImpl {
+public:
+    explicit RawSocket(uint16_t protocol = ETH_P_ALL)
+        : SocketAbstractImpl(socket(AF_PACKET, SOCK_RAW, htons(protocol)))
+    {
+    }
+};
+
+class UdpSocket : public SocketAbstractImpl {
+public:
+    UdpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)) {}
+};
+
+class TcpSocket : public SocketAbstractImpl {
+public:
+    TcpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_STREAM, IPPROTO_IP)) {}
+};
+
+class UdsSocket : public SocketAbstractImpl {
+public:
+    UdsSocket() : SocketAbstractImpl(socket(AF_UNIX, SOCK_STREAM, 0)) {}
```

Comment:
SOCK_SEQPACKET or add an intermediate class `StreamSocket` to share code of `TcpSocket` and `UdsSocket`.

Reply:
If we end up using SOCK_SEQPACKET, we'll derive a new class from SocketAbstractImpl